### ISSUE TITLE
feat: support differential patches on iOS

### ISSUE THREAD

**Issue by eseidel** [shorebirdtech/shorebird#675]
Currently `shorebird ios patch` sends down the entire new Dart binary instead of a diff like Android does.  This means that while Android might send down a patch which is only a few bytes or kilobytes, iOS will always ship down the whole dart section of the app.  Which means the patch size scales with the size of the app rather than the size of the change.

The net effect of this is that during the flutter counter demo of Shorebird, the patch for changing the material theme is 100 bytes on Android, and 4mb on iOS.

This won't be particularly hard to fix. We just wrote it this way because Dart happens to include an ELF parser in its runtime and does not include a Mach-O parser.   We just need to write a Mach-O parser for Dart and then we should be able to use the same diffing/patching strategy that we do on Android.

I estimate this would probably take only a few days for us to fix, we've just had other higher priority requests to service first.

**Comment by eseidel** [comment]
I expect this requires https://github.com/shorebirdtech/shorebird/issues/637

**Comment by eseidel** [comment]
I think the question in my mind with this is should we do a "stop gap" measure of just diffing the elf file to the dylib, or should we go "whole hog" and remove the elf file entirely by teaching Dart how to read the sections it wants out of a mach-o (without calling dlopen)?  I suspect adding a mach-o reader to dart will be pretty straightforward.  But it's also possible the stop-gap will be good enough we shouldn't bother.

**Comment by eseidel** [comment]
I suspect chat gpt might do a decent job of explaining how to write a mach-o reader.

The test case to pass is a small C/c++ file which does a dlopen of a compiled Mac or iOS snapshot and then compares the bytes of the looked up Dart symbols with ones read from ones own custom mach-o reader.

```
const char* DartSnapshot::kVMDataSymbol = "kDartVmSnapshotData";
const char* DartSnapshot::kVMInstructionsSymbol = "kDartVmSnapshotInstructions";
const char* DartSnapshot::kIsolateDataSymbol = "kDartIsolateSnapshotData";
const char* DartSnapshot::kIsolateInstructionsSymbol =
    "kDartIsolateSnapshotInstructions";
```

**Comment by eseidel** [comment]
The attraction of writing our own mach-o reader is that it would remove our hacks to generate an elf file for iOS, which would thus remove possibility of errors like https://github.com/shorebirdtech/shorebird/issues/1228 or https://github.com/shorebirdtech/shorebird/issues/1003

**Comment by eseidel** [comment]
Regardless of what we do (whether we write our own mach-o reader or we just add diffing to our current elf-file setup on iOS) moving to diffing blocks iOS and thus 1.0 as we can't currently have large-apps vending 4mb patches. 🤣

**Comment by eseidel** [comment]
Moving to mach-o would probably replace this block:
https://github.com/shorebirdtech/engine/blob/shorebird/dev/runtime/dart_snapshot.cc#L61
With instead a "readonly" flag to that call, which we'd set to read-only and it would instead use our parser rather than dlopen.

**Comment by eseidel** [comment]
See also https://github.com/shorebirdtech/shorebird/issues/637, I've put mach-o info in both.

**Comment by eseidel** [comment]
So Bryan implemented a solution, which involved diffing an elf library built by Dart on the developer machine (patch) and the base mach-o library built by XCode as part of the IPA.  That worked for Apps, but failed for Frameworks ("add to app" or "hybrid" apps) because the signing information on the Frameworks changed as part of building the final app, and we did the patching at the Framework level.  The root of the problem is that we were diffing containers that contain more than just the 4 files we need.

So I started looking at how we can get these "4 files" to diff with on both the Device and the Host (developer machine).

Device is "easy" in that we can just dlopen/dlsym to look up the files in the mach-o binary, as per normal execution:
https://github.com/shorebirdtech/engine/blob/shorebird/dev/runtime/dart_snapshot.cc#L121

One trick is that dlsym only gives you the start address of the symbol, not the size, but we could easily get the size from the patch, or presumably there is a way to get the size (since Dart itself must figure it out when trying to read the snapshot).

On the host it should also be easy, just a question of tooling.  I asked in the LLVM Discord and got one answer which pointed me to objcopy:
https://discord.com/channels/636084430946959380/636732535434510338/1156268973013352518

objcopy doesn't exist on MacOS (it can be built for MacOS, just not installed by default).  But `segedit` (and older tool) does.

I was able to extract the data and code sections from the macho with this:
```
lipo build/ios/archive/Runner.xcarchive/Products/Applications/Runner.app/Frameworks/App.framework/App -thin arm64 -output arm64.thin
segedit arm64.thin -extract __text __TEXT instructions.o
segedit arm64.thin -extract __const __TEXT data.o
```

I haven't yet tried splitting them, but that is presumably possible by taking the difference in offsets of the two symbols?

I got the offsets from `nm`:
```
% nm build/ios/archive/Runner.xcarchive/Products/Applications/Runner.app/Frameworks/App.framework/App
000000000023b700 S _kDartIsolateSnapshotData
0000000000011640 T _kDartIsolateSnapshotInstructions
0000000000233140 S _kDartVmSnapshotData
0000000000007400 T _kDartVmSnapshotInstructions
```
And then just grepped for them in `otool -l` (which lists segments):
```
eseidel@erics-mbp ios_test % otool -l build/ios/archive/Runner.xcarchive/Products/Applications/Runner.app/Frameworks/App.framework/App | grep 7400 -C5                
   nsects 3
    flags 0x0
Section
  sectname __text
   segname __TEXT
      addr 0x0000000000007400
      size 0x000000000022bd20
    offset 29696
     align 2^6 (64)
    reloff 0
    nreloc 0
eseidel@erics-mbp ios_test % otool -l build/ios/archive/Runner.xcarchive/Products/Applications/Runner.app/Frameworks/App.framework/App | grep 233140  
      addr 0x0000000000233140
eseidel@erics-mbp ios_test % otool -l build/ios/archive/Runner.xcarchive/Products/Applications/Runner.app/Frameworks/App.framework/App | grep 233140 -C5
 reserved1 0
 reserved2 0
Section
  sectname __const
   segname __TEXT
      addr 0x0000000000233140
      size 0x00000000001b4e40
    offset 2306368
     align 2^6 (64)
    reloff 0
    nreloc 0
```

**Comment by eseidel** [comment]
NSGetSectionDataInObjectFileImage may also help, if it exists on iOS.
https://opensource.apple.com/source/cctools/cctools-822/man/NSObjectFileImage.3.auto.html

**Comment by eseidel** [comment]
NSGetSectionDataInObjectFileImage is deprecated / not available.

But we can just dump the whole segments on the Mac, and send diffs to the segments.

**Comment by eseidel** [comment]
OK, so I think the path here is to use the segedit command from above on the host:
```
lipo build/ios/archive/Runner.xcarchive/Products/Applications/Runner.app/Frameworks/App.framework/App -thin arm64 -output arm64.thin
segedit arm64.thin -extract __text __TEXT instructions.o
segedit arm64.thin -extract __const __TEXT data.o
```
The remaining trick we need to figure out is how to split those files one more time into their DartIsolate/DartVM halves.  That's important just to make it easy to load them into the Dart VM on the client.
They can be fed into the Dart VM via these deprecated properites:
https://github.com/shorebirdtech/engine/blob/shorebird/dev/common/settings.h#L110

Then on the client we can just use the symbols (looked up via dsym) as the buffers to reinflate from.  We can separately send down the length of the symbols if needed (since dsym won't tell you).

**Comment by eseidel** [comment]
I suspect that we could try splitting those segments in two, using the symbol offsets from:
```
% nm build/ios/archive/Runner.xcarchive/Products/Applications/Runner.app/Frameworks/App.framework/App
000000000023b700 S _kDartIsolateSnapshotData
0000000000011640 T _kDartIsolateSnapshotInstructions
0000000000233140 S _kDartVmSnapshotData
0000000000007400 T _kDartVmSnapshotInstructions
```
I'm not sure if they will end up padded for alignment at the end though.

**Comment by eseidel** [comment]
After further discussion, going to wait until @kmillikin is further with mixed-mode and see what affects mixed-mode has on our desired patch format.

**Comment by eseidel** [comment]
Turns out Dart has at least 3 (partial) implementations of Mach-O parsing:
https://github.com/dart-lang/sdk/blob/main/runtime/platform/mach_o.h
https://github.com/dart-lang/sdk/blob/main/pkg/dart2native/lib/macho.dart
https://github.com/dart-lang/sdk/blob/main/pkg/native_stack_traces/lib/src/macho.dart

**Comment by eseidel** [comment]
I've implemented a mach-o parser, this should now be possible.  https://github.com/dart-lang/sdk/pull/53953

**Comment by eseidel** [comment]
I believe the way this will work is that we will use our fancy new MachOLoader technology in Dart to dump the 4 sections of the snaphot out into a single stable buffer/file.  We will then use that buffer/file as the "base" off of which to create the differential patch.  And then we'll ship the differential patch down to the devices (just like we do on Android) and then re-constitute the original patch by again on the device creating an identical stable base buffer/file by again using the macho loader built into Dart to read the 4 sections out of the base App binary and then passing that off to the rust code to use as the "base" (instead of trying to use the actual on-disk file like we do on Android).

As I type this I realize that we might not even need the Mach-O loader on the device, since we already have the symbols loaded into memory (as the base snapshot anyway).  We'd just need to know how large they were, which we could even send down to the device if needed.  That said, the MachO loader is easy to use and would tell you the size of the symbols.  You could even use the Mach-O loader in a mode where it just parsed out the symbol sizes for you and didn't try to allocate buffers.

To wire this up might require exposing macho_loader as a gn target, since it's located in bin/ (like elf_loader) rather than vm/ and thus is not linked into the Dart library by default.

We will also probably need to expose new API on the updater library to pass in the generate base buffer/file.  Either a path to one, or a bytes pointer and length.  Might also need to make a small change to the patch library to support reading from a buffer rather than just a file as well if it doesn't already support that (which it might?).

Separately we can also remove the vm_data and vm_istructions from the patch snapshot (and this "base" file since we've learned those are no longer needed in the patch since they're stable between all dart programs from the same version of Dart.

**Comment by bryanoltman** [comment]
@eseidel Having a little trouble figuring out how we get the length of each of the four sections from what the Mach-O Loader spits out (the four `uint8_t**` values). Don't want to interrupt the current debugging session, but when you have a sec :)

**Comment by eseidel** [comment]
I don't think mach-o includes length of symbols, but it does include size of sections.  So with the size of the sections and the two symbols I think you can compute:
vm_size = isolate_ptr - vm_ptr
isolate_size = section_size - isolate_ptr?

Or just dump the two sections, not even worrying about the sizes, e.g, just __text__ and __data__ concatenated.

**Comment by eseidel** [comment]
Bryan has this fixed locally.  We plan to release the fixes this week.

**Comment by eseidel** [comment]
This just shipped in 0.22.0!  Patch sizes should now be 10x smaller and grow based on the amount of Dart changed, rather than the total amount of Dart in the app.
